### ISSUE TITLE
Hard crash, when replacing VHD-based drives in a pool, during scrubbing

### ISSUE THREAD

**Issue by vmlemon** [openzfsonwindows/ZFSin#151]
Using OpenZFS 20190612, and Windows 10 Version 1903 Build 18362.30

Unfortunately, I couldn't get a MiniDump (in fact, since the system locks up, hard, thanks to memory corruption, I don't even get a BSOD!), but I found a really horrible scenario, that leads to an amusing failure mode, on my machine (which only has 4GB of RAM installed, and is non-upgradeable):

![DSC_0034_Resize](https://user-images.githubusercontent.com/35642/60378916-0120c080-9a22-11e9-8e67-9473695e097f.JPG)

Roughly... 

 * Create a 100GB VHD (the files are sparse, but it's best to use dynamically-expanding, rather than fixed-size, for this exercise), using Disk Management, on an NTFS volume, and <code>zpool create -d -O dedup=on -O compress=on -O normalization=formD Store2 PHYSICALDRIVEx</code> (where <code>x</code> is the Disk Management number, for the new virtual disk)
 * Upgrade the new pool, to the latest feature set, and fill the new dataset, with ~11GB of data
 * Create a second, 200GB VHD, using Disk Management, also on an NTFS volume
 * Start a scrub of the pool, and keep it running
 * Use <code>zpool add Store2 PHYSICALDRIVEy</code>, where <code>y</code> corresponds to the new, 200GB disk's number, in Disk Management
 * Enable auto-expansion, on the pool, with <code>zpool set autoexpand=on Store2</code>
 * Attempt to replace the 100GB disk, with the 200GB disk, using <code>zpool replace Store2 /dev/physicaldriveX /dev/physicaldriveY</code> (<code>X</code> = 100GB VHD; <code>Y</code> = 200GB VHD)
 * After a while, <code>zpool status</code> should say that it's evacuating the contents of the 100GB disk, to the 200GB disk, whilst the scrubbing is still running...

Near the end of the process, the entire system will crash, with either graphical corruption, or a BSOD; and, if you attempt to reattach the VHDs, after a reboot, and do a <code>zpool import -af</code>, the OS will lock up, shortly after importing the pool.

**Comment by lundman** [comment]
Woo exciting!  The first thing that springs to mind is that on OsX the scrub code will eat the stack in anything build debug, ie, needs -O. So that could be something to look at, I've not checked the stack size in Windows, but could be Release build will work better.

**Comment by lundman** [comment]
kernel-mode stack is 24 KB.  Compared to OsX's 16KB. So I'm surprised that the larger stack would die the same way - well, we should check that out either way

**Comment by vmlemon** [comment]
Aah, interesting. Just remembered that I might have used "zfs detach", instead of "zfs replace", since it was a pain to figure out just which device node style, and replacement technique, I had to use. 

(I've been fortunate, in that I've never had to replace a disk in a ZPool, but in this case, I was doing some preparation, for making an image of a 128GB disk, and needed some extra headroom, in case compression didn't quite account for the final file size, and the process failed).

**Comment by vmlemon** [comment]
Looks like I can move the 2nd VHD to my 4TB HDD, but if I try to move the 1st one, it causes a hang. Subsequently, after rebooting, taking a snapshot, and overwriting the 1st VHD, I receive a "KERNEL SECURITY CHECK FAILURE" BSOD.

**Comment by lundman** [comment]
Ok, reading the instructions:
```
* Use zpool add Store2 PHYSICALDRIVEy, where y corresponds to the new, 200GB disk's number, in Disk Management
* Attempt to replace the 100GB disk, with the 200GB disk, using zpool replace Store2 /dev/physicaldriveX /dev/physicaldriveY (X = 100GB VHD; Y = 200GB VHD)
```

If we are talking about 1x 100GB, and 1x 200GB disk, and you tried to replace the 100GB with the "already active 200GB" disk (from the `zpool add` command) the problem is actually that it should definitely stop you. The disks _should_ be opened exclusive, and it should error out. If it does not, then we have two opens to the same disk and I can imagine that it crashes badly.

**Comment by lundman** [comment]
Hmm no, can't replicate crash here - the kernel should not open disks with share_write on, but it doesn't even get that far here.

**Comment by vmlemon** [comment]
Interesting. 

Not sure if "<code>zpool detach</code>"-ing (vs <code>replace</code>-ing) the 100GB disk is what makes the difference (it took me a while, to find any way of replacing a disk, that even pretends to work, in fairness), although once I've moved off the old VHDs, and my debris from trying to recover data from an ExFAT-formatted memory card (looks like I'm having to use an OEM version of Paragon's NTFS driver, under Mac OS X, since I sometimes get BSODs, moving those files, from NTFS, to ZFS, under Windows), I'll probably have another go at trying to trigger the bug, with a fresh pool...

The guide was done purely from memory, as a post-mortem of the steps that I took, before it crashing, that said - so if there's any missing steps, or differing commands, I wouldn't be surprised. :(

**Comment by lundman** [comment]
You can't `zpool detach` if you did `zpool add`. But you can if it was `zpool attach`. I'm going to check if scrub/resilver will overflow the stack on Windows in debug builds.

**Comment by vmlemon** [comment]
Gotcha - just had a look at https://docs.oracle.com/cd/E53394_01/html/E54801/ghzvz.html, and it seems that it's definitely the "attach/detach" variant of the process, that I did, now that I think of it.

Once the "Store2.vhd" has finished transferring, I'll plug the disks into a VM, to see what happens, if I try to import the pool into another ZFS implementation, and extract the ZPool history.

FWIW, it could be handy to add the ability to use the "regular" PHYSICALDRIVEx syntax, to the add/remove, and attach/detach commands, or a note stating that you've got to do "/dev/physicaldriveX", with quotes, in order to access a raw disk, if it doesn't require tearing everything apart, to do so...

**Comment by vmlemon** [comment]
I should really test it, with both approaches, and see if it's still possible to reproduce the crash, on my end, though.

**Comment by lundman** [comment]
There is some work to do with the path thing for sure, the best way to refer to existing devices though, is to use guids. Ie, use `zpool status -g` to list devices by guids, then use the guid with the command to replace/detach etc. Like `zpool replace tank 423132132531 PHYSICALDRIVE2`.

Eventually, we'll find all the places to handle `/dev/physicaldriveX` and `PHYSCIALDRIVEx` shortcuts.

**Comment by vmlemon** [comment]
Looks like I can take those two VHDs, and import them, under an OpenIndiana Hipster VM, and <code>zpool status</code> claims that the pool is running completely on the Store3 VHD, with all of the data migrated, and the scrub is still running:

![VirtualBox_MidnightBSD_03_07_2019_13_51_59](https://user-images.githubusercontent.com/35642/60593408-97dedb80-9d9a-11e9-97bf-ec0953303313.png)

Obviously doesn't help us fix the Windows implementation, to ensure that it doesn't happen again, in the future, but at least it provides reassurance that my data isn't completely unrecoverable, even if the pool was initially only created as an experiment.

Since we can now dig into the pool history, it's pretty much:
<code>
[Pool birth command is missing - Created under Windows]
zfs snapshot Store2@Populated
[Series of snapshot commands]
zpool add Store2 PHYSICALDRIVE5
zpool set autoexpand=on Store2
zpool resilver Store2
zpool scrub Store2 [Started on Windows, continues to run, on OpenIndiana...]
zpool remove Store2 /dev/physicaldrive3 [Windows]
zpool import -af [Pool reawakens, under OpenIndiana]
</code>
If the scrub ever completes, I'll try reimporting it, under Windows, to see what happens.

**Comment by vmlemon** [comment]
Yup, so you "add" the 200GB device, autoexpand the pool, and then start a scrub, before kicking the ladder out from underneath everything, by "remove"-ing the 100GB pool, and wait for the two processes to get into perfect flow, before it eventually crashes...

This is on the debug build of ZFSin, and not the release one.

**Comment by vmlemon** [comment]
(The evacuation process begins to run, and starts copying the data from the 100GB disk, onto the 200GB disk, after the "remove", and towards the end of the evacuation process, the crash happens, whilst the scrub continues)

**Comment by vmlemon** [comment]
Aha! So, after mounting the Store3 VHD, I receive an instant crash, with graphical corruption, with the debug build.

Looks like it's the act of scrubbing, itself that causes memory corruption.
